LSM: SuperBlockFreeSet encode_with_staging

sentientwaffle opened the pull request:
If we crash without writing the `SuperBlockFreeSet` with the staged blocks, they will leak. Add `SuperBlockFreeSet.encode_with_staging` which encodes staged blocks as free.

sentientwaffle committed:
LSM: SuperBlockFreeSet encode_with_staging

If we crash without writing the `SuperBlockFreeSet` with the staged blocks, they will leak. Add `SuperBlockFreeSet.encode_with_staging` which encodes staged blocks as free.

sentientwaffle committed:
LSM: SuperBlockFreeSet.last_acquired

`SuperBlockFreeSet.last_acquired` can be used to help verify different replicas' storage is deterministic (in sync).

koekiebox reviewed (approved):
Looks good to me.

ifreund commented on src/lsm/superblock_free_set.zig:
```diff
@@ -72,17 +72,30 @@ pub const SuperBlockFreeSet = struct {
 
-    // TODO Add unit tests for count_released() and count_acquired().
+    /// Returns the address of the highest allocated block.
+    pub fn last_acquired(set: SuperBlockFreeSet) ?u64 {
```
I think this function name could be improved. If I understand correctly, the highest allocated block isn't necessarily the most recently acquired block as a block with a lower address may have been free'd and re-acquired since the highest block was acquired.

I realize now after writing this that you probably meant "last" as in "physically last in order on disk" instead of "most recently acquired" as I read it, I still think it would be worth resolving this ambiguity though.

sentientwaffle replied:
That is a good point — the idea is to provide a (very rough) estimate of the amount of disk used, so the relevant "last" block has the highest address, but is not necessarily last chronologically.

jorangreef replied:
@sentientwaffle to borrow from your comment, what about `highest_allocated_block_address()`?

jorangreef commented on src/lsm/superblock_free_set.zig:
```diff
@@ -72,17 +72,30 @@ pub const SuperBlockFreeSet = struct {
+            const address = block + 1;
+            return address;
+        } else {
+            return null;
```
We could assert before returning null that there are zero acquired blocks in the set.

jorangreef commented on src/lsm/superblock_free_set.zig:
```diff
@@ -175,12 +188,32 @@ pub const SuperBlockFreeSet = struct {
     }
 
     /// Returns the number of bytes written to `target`.
+    /// The encoded data does *not* incldue staged changes.
```
```suggestion
    /// The encoded data does *not* include staged changes.
```

jorangreef commented on src/lsm/superblock_free_set.zig:
```diff
@@ -175,12 +188,32 @@ pub const SuperBlockFreeSet = struct {
+    pub fn encode_with_staging(set: *SuperBlockFreeSet, target: []align(@alignOf(usize)) u8) usize {
+        const count_free = set.count_released();
+        const count_staged = set.staging.count();
+        assert(target.len == SuperBlockFreeSet.encode_size_max(set.blocks.bit_length));
```
This assert on the target arg could be moved earlier right up to the top of the function, since it's a pre-condition assert and doesn't depend on `count_free/staged` I think.

jorangreef replied:
It would bring the counts also a fraction closer to where we use them.

jorangreef commented on src/lsm/superblock_free_set.zig:
```diff
@@ -175,12 +188,32 @@ pub const SuperBlockFreeSet = struct {
+
+        // Temporarily mark the staged blocks as free.
+        set.blocks.toggleSet(set.staging);
+        assert(set.blocks.count() == count_free + count_staged);
```
Nice assertion to have.

jorangreef replied:
Since staging is so powerful here, with the ability to toggle all bits if staging is accidentally initialized the wrong way, it would be good to assert in `init()` that `staging.count() == 0`.

jorangreef commented on src/lsm/superblock_free_set.zig:
```diff
@@ -175,12 +188,32 @@ pub const SuperBlockFreeSet = struct {
+        const encode_size = ewah.encode(bitset_masks(set.blocks), target);
+        // Restore the changes: mark the staged blocks as allocated again.
+        set.blocks.toggleSet(set.staging);
+        assert(set.blocks.count() == count_free);
```
If you like, what do you think about putting this in a defer block right where we do the toggle? Then the return line for this function can match that of `encode()`.

jorangreef reviewed (approved):
Thanks DJ, looks great.

sentientwaffle committed:
LSM: SuperBlockFreeSet review cleanup

- Assert initial bitset state.
- Rename `last_acquired` to `highest_allocated_block_address` to clarify that it returns the last block on disk, not the last one written chronologically.
- Misc cleanup.
